torch/_higher_order_ops/associative_scan.py

```diff
@@ -40,10 +41,39 @@ def wrap_combine_fn_flat(*args, combine_fn, spec, num_leaves):
     return combined_leaves
 
 
+def check_args(input, combine_fn, leaves, tree, dim):
+    assert len(leaves) >= 1, "expected at least 1 input leaf"
+    assert all(
+        isinstance(x, torch.Tensor) for x in leaves
+    ), "input leaves must be a Tensor"
+    shape = leaves[0].shape
+    ndim = len(shape)
+    dim = utils.canonicalize_dim(ndim, dim)
+
+    for x in leaves[1:]:
+        assert x.shape == shape, "All input tensors must have the same shape"
+
+    out = combine_fn(
+        pytree.tree_unflatten(
+            [e[slice_along_axis(0, 1, stride=None, dim=0)] for e in leaves], tree
+        ),
+        pytree.tree_unflatten(
+            [e[slice_along_axis(0, 1, stride=None, dim=0)] for e in leaves], tree
+        ),
+    )
+
+    out_leaves, tree_out = pytree.tree_flatten(out)
+    assert (
+        tree == tree_out
+    ), "The pytree of the output of the operator needs to match the input pytree"
+
+
 def associative_scan(
     combine_fn: Callable[[pytree.PyTree, pytree.PyTree], pytree.PyTree],
     input: pytree.PyTree,
     dim: int,
+    reverse: bool = False,
+    host_side: bool = False,
```

Comment:
why do we want to have a flag, rather than dispatching using the devices in the input. We could return a function that does
```python
def associative_scan(input, ...):
  if input.is_cpu: ...
  else: ...
```
and have `torch.compile`  trace appropriately.

Reply:
I think that there might be a confusion here. The _host-side_ here does not refer to CPU-only in this case. Rather it refers to the fact that no _device-side_ operations, like the `tl.associative_scan` are invoked. I added a testcase to check for the compatibility for CUDA as well. 

Interestingly to me though, the CUDA test fails for me, not because of the device, but rather because when after using the `reverse=False` flag and compiling the function, the reverse is set to `True` and invoked again, it doesn't issue a new compilation and hence still uses the `reverse=False `during execution, which then fails the test. This only happens on the CUDA device, not on the CPU. Do you have any suggestions as to how I can resolve this?

Reply:
Oh, I see. Then it'd be nice to bikeshed the name of the parameter a bit...

Can you give a small repro for the issue?

Reply:
What about if we reverse the meaning and have a parameter like `use_explicit_tl` that would default to `True` to not change the original behavior?

Of course. This code snippet can reproduce the error that I am seeing. If one sets the device to `torch.device('cpu')` it will work, if one sets the device to `torch.device('cuda')`, it will error out.
```
def fct(x: torch.Tensor, y: torch.Tensor):
    return x

x = torch.randn(3, 2, 2, requires_grad=True, device=torch.device('cuda'))
result = associative_scan(fct, x, 0, host_side=True, reverse=False)
expexted_result = _fake_associative_scan(fct, x, 0, reverse=False)
self.assertEqual(result, expexted_result)
result = associative_scan(fct, x, 0, host_side=True, reverse=True)
expexted_result = _fake_associative_scan(fct, x, 0, reverse=True)
self.assertEqual(result, expexted_result)
```

Reply:
For the keyword, what about something like `generic_scan=False` as the default, and writing a good documentation explaining the pros and cons (capabilities vs. performance) of using this flag?

About that issue... I don't know off the top of my head. The only thing that comes to mind is that somehow `torch.compile` may not be caching that flag and that's why it's not recompiling it? But I don't see a priori why would that be the case.

Reply:
Right, to me this sounds good. I have updated the code accordingly.

Regarding this issue, I was wrong in my initial assessment. After digging into this issue, I found that the function behaves differently when using `torch.compile` or not. For example, this code
```
def fct(x: torch.Tensor, y: torch.Tensor):
    return x + y

for n in range(15):
    x = torch.arange(n, device=torch.device('cuda'))
    associative_scan1 = torch.compile(associative_scan)
    associative_scan2 = associative_scan
    result1 = associative_scan1(fct, x, 0, True, True)
    result2 = associative_scan2(fct, x, 0, True, True)
    print(str(n), ': ', ['{:2d}'.format(int(r.cpu().numpy())) for r in result1])
    print(str(n), ': ', ['{:2d}'.format(int(r.cpu().numpy())) for r in result2])
    print(str(n), ': ', [' T' if result1[i] == result2[i] else ' F' for i in range(n)])
    print('-'*80)
```
yields 
```
0 :  []
0 :  []
0 :  []
--------------------------------------------------------------------------------
1 :  [' 0']
1 :  [' 0']
1 :  [' T']
--------------------------------------------------------------------------------
2 :  [' 1', ' 1']
2 :  [' 1', ' 1']
2 :  [' T', ' T']
--------------------------------------------------------------------------------
3 :  [' 3', ' 3', ' 2']
3 :  [' 3', ' 3', ' 2']
3 :  [' T', ' T', ' T']
--------------------------------------------------------------------------------
4 :  [' 6', ' 6', ' 6', ' 3']
4 :  [' 6', ' 6', ' 5', ' 3']
4 :  [' T', ' T', ' F', ' T']
--------------------------------------------------------------------------------
5 :  ['10', '10', ' 9', ' 7', ' 4']
5 :  ['10', '10', ' 9', ' 7', ' 4']
5 :  [' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
6 :  ['15', '15', '14', '12', ' 9', ' 5']
6 :  ['15', '15', '14', '12', ' 9', ' 5']
6 :  [' T', ' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
7 :  [' 0', ' 0', '21', '21', '18', '18', ' 6']
7 :  ['21', '21', '20', '18', '15', '11', ' 6']
7 :  [' F', ' F', ' F', ' F', ' F', ' F', ' T']
--------------------------------------------------------------------------------
W0625 20:25:33.179000 139648747538240 torch/_dynamo/convert_frame.py:762] [0/8] torch._dynamo hit config.cache_size_limit (8)
W0625 20:25:33.179000 139648747538240 torch/_dynamo/convert_frame.py:762] [0/8]    function: 'inner' (/data_malta3_ssd/pytorch_git/torch/_dynamo/external_utils.py:36)
W0625 20:25:33.179000 139648747538240 torch/_dynamo/convert_frame.py:762] [0/8]    last reason: tensor 'L['args'][1]' size mismatch at index 0. expected 0, actual 8
W0625 20:25:33.179000 139648747538240 torch/_dynamo/convert_frame.py:762] [0/8] To log all recompilation reasons, use TORCH_LOGS="recompiles".
W0625 20:25:33.179000 139648747538240 torch/_dynamo/convert_frame.py:762] [0/8] To diagnose recompilation issues, see https://pytorch.org/docs/main/torch.compiler_troubleshooting.html.
8 :  ['28', '28', '27', '25', '22', '18', '13', ' 7']
8 :  ['28', '28', '27', '25', '22', '18', '13', ' 7']
8 :  [' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
9 :  ['36', '36', '35', '33', '30', '26', '21', '15', ' 8']
9 :  ['36', '36', '35', '33', '30', '26', '21', '15', ' 8']
9 :  [' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
10 :  ['45', '45', '44', '42', '39', '35', '30', '24', '17', ' 9']
10 :  ['45', '45', '44', '42', '39', '35', '30', '24', '17', ' 9']
10 :  [' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
11 :  ['55', '55', '54', '52', '49', '45', '40', '34', '27', '19', '10']
11 :  ['55', '55', '54', '52', '49', '45', '40', '34', '27', '19', '10']
11 :  [' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
12 :  ['66', '66', '65', '63', '60', '56', '51', '45', '38', '30', '21', '11']
12 :  ['66', '66', '65', '63', '60', '56', '51', '45', '38', '30', '21', '11']
12 :  [' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
13 :  ['78', '78', '77', '75', '72', '68', '63', '57', '50', '42', '33', '23', '12']
13 :  ['78', '78', '77', '75', '72', '68', '63', '57', '50', '42', '33', '23', '12']
13 :  [' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
14 :  ['91', '91', '90', '88', '85', '81', '76', '70', '63', '55', '46', '36', '25', '13']
14 :  ['91', '91', '90', '88', '85', '81', '76', '70', '63', '55', '46', '36', '25', '13']
14 :  [' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T']
-------------------------------------------------------------------------------- 
```
I tried to track it down further, but unfortunately I have no clue where this could originate from.